### DISCUSSION TITLE
tremor send all unsed css to front end

### DISCUSSION THREAD

**Discussion by techUdayMungalpara** [tremorlabs/tremor-npm#1011] [General]
The tremor library should implement a process to identify and remove any unused CSS classes before sending the CSS to the frontend. This can help reduce the overall file size and optimize the performance of the application.

**Comment by marlatte** [comment]
It looks like this is due to all the patterns being included in the `safelist` in `tailwind.config.js`. [Tailwind's docs](https://tailwindcss.com/docs/content-configuration#safelisting-classes) say: 

> Safelisting is a last-resort, and should only be used in situations where it’s impossible to scan certain content for class names. These situations are rare, and you should almost never need this feature.

I can't find anything in the Tremor docs or here on GH where it's explained why it needs to force Tailwind generate all these unused classes. If I'm reading the `safelist`'s regex right, the patterns are creating...
```
(3 patterns * 22 colors * 11 shades) + (3 patterns * 22 colors * 11 shades * 2 variants) = 2,178 total classes
```

**Reply by severinlandolt** [reply]
Hi! Yes, we need this because we dynamically generate gradients and backgrounds for the charts. With our copy-and-paste components, we eliminated this issue by having dedicated chartUtils: https://tremor.so/docs/utilities/chartUtils#add-the-following-code-to-your-object-object-file.

We are trying to get off the safelist in v4.
